plan/exhaust_physical_plans.go

```diff
@@ -278,36 +278,36 @@ func (p *LogicalJoin) getHashJoin(prop *requiredProp, innerIdx int) *PhysicalHas
 	return hashJoin
 }
 
-// joinKeysMatchIndex checks if all keys match columns in index.
-// It returns a slice a[] what a[i] means keys[i] is related with indexCols[a[i]].
+// joinKeysMatchIndex checks whether the join key is in the index.
+// It returns a slice a[] what a[i] means keys[i] is related with indexCols[a[i]], -1 for no matching column.
+// It will return nil if there's no column that matches index.
 func joinKeysMatchIndex(keys, indexCols []*expression.Column, colLengths []int) []int {
-	if len(indexCols) < len(keys) {
-		return nil
-	}
 	keyOff2IdxOff := make([]int, len(keys))
-	for keyOff, key := range keys {
-		idxOff := joinKeyMatchIndexCol(key, indexCols, colLengths)
-		if idxOff == -1 {
-			return nil
-		}
-		keyOff2IdxOff[keyOff] = idxOff
-	}
-	return keyOff2IdxOff
-}
-
-func joinKeyMatchIndexCol(key *expression.Column, indexCols []*expression.Column, colLengths []int) int {
-	for idxOff, idxCol := range indexCols {
-		if colLengths[idxOff] != types.UnspecifiedLength {
+	for i := range keyOff2IdxOff {
+		keyOff2IdxOff[i] = -1
+	}
+	// There should be at least one column in join keys which can match the index's column.
+	matched := false
+	tmpSchema := expression.NewSchema(keys...)
+	for i, idxCol := range indexCols {
```

Comment:
We could break early in this loop. We only consider the situation that index prefix matches the join key.

Reply:
Actually we cannot. We need to deal with this case: `t1 join t2 on t1.b=t2.b where t1.a=1 and t1.c=1`, if there's index on t1's (a, b, c). We can use index join. But t1.b is not the prefix of the index. If we break rather than continue, we will not choose index join.